src/Compilers/CSharp/Portable/Lowering/LocalRewriter/DelegateCacheRewriter.cs

```diff
@@ -0,0 +1,212 @@
+
+    private static bool TryGetOwnerFunction(MethodSymbol currentFunction, TypeSymbol delegateType, MethodSymbol targetMethod, [NotNullWhen(true)] out MethodSymbol? ownerFunction)
+    {
+        if (targetMethod.MethodKind == MethodKind.LocalFunction)
+        {
+            // Local functions can use type parameters from their enclosing methods!
+            //
+            // For example:
+            //   void Test<T>()
+            //   {
+            //       var t = Target<int>;
+            //       static object Target<V>() => default(T);
+            //   }
+            //
+            // Therefore, without too much analysis, we select the closest generic enclosing function as the cache container owner.
+
+            for (Symbol? enclosingSymbol = currentFunction; enclosingSymbol is MethodSymbol enclosingMethod; enclosingSymbol = enclosingSymbol.ContainingSymbol)
+            {
+                if (enclosingMethod.Arity > 0)
+                {
+                    ownerFunction = enclosingMethod;
+                    return true;
+                }
+            }
+
+            ownerFunction = null;
+            return false;
+        }
+
+        var usedTypeParameters = PooledHashSet<TypeParameterSymbol>.GetInstance();
+        try
+        {
+            FindTypeParameters(delegateType, usedTypeParameters);
+            FindTypeParameters(targetMethod, usedTypeParameters);
+
+            for (Symbol? enclosingSymbol = currentFunction; enclosingSymbol is MethodSymbol enclosingMethod; enclosingSymbol = enclosingSymbol.ContainingSymbol)
+            {
+                if (usedTypeParametersContains(usedTypeParameters, enclosingMethod.TypeParameters))
+                {
+                    ownerFunction = enclosingMethod;
+                    return true;
+                }
+            }
+
+            ownerFunction = null;
+            return false;
+        }
+        finally
+        {
+            usedTypeParameters.Free();
+        }
+
+        static bool usedTypeParametersContains(HashSet<TypeParameterSymbol> used, ImmutableArray<TypeParameterSymbol> typeParameters)
+        {
+            foreach (var typeParameter in typeParameters)
+            {
+                if (used.Contains(typeParameter))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
+
+    private static void FindTypeParameters(TypeSymbol type, HashSet<TypeParameterSymbol> result)
+        => type.VisitType(TypeParameterSymbolCollector, result, visitCustomModifiers: true);
+
+    private static void FindTypeParameters(MethodSymbol method, HashSet<TypeParameterSymbol> result)
+    {
+        FindTypeParameters(method.ContainingType, result);
+
+        foreach (var typeArgument in method.TypeArgumentsWithAnnotations)
+        {
+            typeArgument.VisitType(type: null, typeWithAnnotationsPredicate: null, TypeParameterSymbolCollector, result, visitCustomModifiers: true);
```

Comment:
>TypeParameterSymbolCollector [](http://example.com/codeflow?start=83&length=28)

Are we creating this delegate over and over again?

Reply:
This is an interesting question. We're making the compiler not creating delegates again and again. If the toolset compiler is using C# 10, it would produce non-caching code here, but since the produced compiler will be using C# 10 again as toolset, Code here is not really reachable. If the toolset compiler is using C# preview, then the first produced compiler(without this change) will produce non-caching code here, but then the next compiler produced by the first one will produce caching code here.

What makes this more interesting is that, we are in a loop here, @VSadov mentioned that it would be lovely we produce local caching code without branching, probably in a separate PR.

Reply:
I suggest caching the delegate in a static readonly field as we do that for other common usages of VisitType.

Reply:
OK.

Reply:
Changed. Please check 49c7420.